### PR TITLE
Porting the current command language on PLY

### PR DESCRIPTION
**Pull request by kajojify**
Hello! This is my first GSoC PR :tada: 
Sorry for the delay - I spent much time on [thinking about the language](https://github.com/mitmproxy/mitmproxy/issues/3087#issuecomment-386811558), asking @cortesi questions and [playing with PLY](https://github.com/mitmproxy/mitmproxy/issues/3159). This PR transfers the current version of command language to PLY, adapting command features and preparing to have new command language syntax.

Henceforth, I will try to keep my contributions pace high.
### Step 1. Lexing.

So the sequence of actions was:
_Define the places, where we lex anything_ -> _see the interface_ -> _write the lexer according to PLY docs and available examples_ -> _adapt it to the current implementation_ -> _replace the current lexer with the new lexer_ -> _check broken things_.
- **Places.** The current [lexer](https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/command.py#L25) occurs 2 times: in `execute` [method](https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/command.py#L235) and in `parse_partial` [method](https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/command.py#L153).
- **Interface.** Changed in comparison to the current implementation: 
_Creating lexer:_
```
from mitmproxy.language import lexer
lexer = lexer.create_lexer("flow.mark.toggle @focus")
```
Created lexer supports the iteration protocol and also has `token()` method to give you next token from lexed string.

_Getting_ tokens:
```
from mitmproxy.language import lexer
tokens = lexer.get_tokens("flow.mark.toggle @focus")
```
or
```
from mitmproxy.language import lexer
lexer = lexer.create_lexer("flow.mark.toggle @focus")
tokens = [token.value for token in lexer]
```
- **New lexer.** The whole lexer class is built according to [PLY documentation](http://www.dabeaz.com/ply/ply.html#ply_nn3). Separate `build` method is a recommendation from docs. 

- **Broken things.** I know, that some command don't work now and `view.load"mypath"` will be lexed like `view.load` and `"mypath"`, which is incorrect. I still fix all these things. I know this, but my main goal is adapting.

My next steps:
- get rid of lexing bugs;
- implement parser. It is almost ready.

### CODE REVIEW

**Comment by mhils** [comment]
This looks like a very reasonable change! 👍 
We also need to convince CI, but other than that I'm good with this. 😃

**Comment by cortesi** [comment]
Thumbs up on this approach. Once we have tests passing, we should merge so we can keep moving on.

File: setup.py [hunk 1]
```diff
@@ -70,6 +70,7 @@
         "kaitaistruct>=0.7,<0.9",
         "ldap3>=2.5,<2.6",
         "passlib>=1.6.5, <1.8",
+        "ply>=3.4, <3.12",
```
Comment (Kriechi) [inline setup.py]: Did you actually test this with ply 3.4?
That release is pretty old (2011), so I think we can bump this to at least 3.6 (2015) or even 3.10 (2017)...

**Comment by kajojify** [comment]
### Step 2. Parsing.
According to the wiki page:

> **Parsing** (syntax analysis) is the process of analysing a string of symbols, either in natural language, computer languages or data structures, conforming to the rules of a formal grammar.

**What do we need?** 
_get user input_ -> _lex it_ -> _juxtapose the tokens with formal language grammar_ -> _execute_
The first two steps are done. We get user input, [create a lexer](https://github.com/kajojify/mitmproxy/blob/language/mitmproxy/language/lexer.py#L55) basing on the user input and pass the lexer to the parser. So, now it's time for parser to show itself. 
Before writing the code, we need to describe the formal grammar of our language using BNF notation.

**Minimized BNF notation**, which takes actual tokens into account :

```
command_line : empty 
             | command_call
command_call : COMMAND argument_list"
argument_list : empty
              | argument_list argument
argument : PLAINT_STR
         | QUOTED_STR
         | COMMAND
empty : ""
```
Now let's join BNF notation with the code.

**Parser class.** It was written according to the [PLY docs](http://www.dabeaz.com/ply/ply.html#ply_nn23). In short, every BNF rule has the corresponding method, which describes what to do:
> The semantic behavior of a language is often specified using a technique known as syntax directed translation. In syntax directed translation, attributes are attached to each symbol in a given grammar rule along with an action. Whenever a particular grammar rule is recognized, the action describes what to do.

The main idea of the parser is to pass command manager into parser class as an `__init__` argument and then use `call_strings` method from within, when defining COMMAND token and its argument list. 
`command_manager.call_strings(COMMAND, argument_list)`

**Parser interface.**
_Parser creating:_
```
from mitmproxy.language import parser

command_parser = parser.create_parser(command_manager)
```
_Parsing_
```
from mitmproxy.language import lexer

lexer = lexer.create_lexer("flow.mark.toggle @focus")
command_parser.parse(lexer)
```

**Comment by kajojify** [comment]
### Step 3. Additional features.
Lexer and parser are implemented, but we still have to do something with our interactive features: **syntax highlighting**, **autocompletion**, **underlying type conversion**.
In fact, all these features work like they worked before, except one thing.
Now you can use arbitrary number of white spaces between your tokens and this won't break anything:
image

The lexer has two states: _INITIAL_ and _interactive_. _INITIAL_ state doesn't take into account white spaces and is needed for parsing and execution. _interactive_ - takes into account white spaces and is needed for command prompt features.
When we use `execute` method, we use _INITIAL_ state. When we use `parse_partial`, we use _interactive_ state. So, for example `lexer.get_tokens("cut.clip @focus request.host ")` will give us such list inside `parse_partial`:`['cut.clip', ' ', '@focus', 'request.host', ' ']`, but if to use it with `state=INITIAL` param, it will return `['cut.clip', '@focus', 'request.host']`. It was done in order to take control over tokens separation on command prompt. 

The only thing I still need is to deal with test coverage of lexer and parser. The other tests seem to be fixed.

**Comment by cortesi** [comment]
Excellent progress. Let's fix the coverage so we can merge! :)

**Comment by Kriechi** [comment]
Sorry, I didn't mean to close the PR or the language branch.
Would you mind pushing it again?

**Pull request merged**